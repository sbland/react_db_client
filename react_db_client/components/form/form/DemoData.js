import { filterTypes } from '@react_db_client/constants.client-types';

export const demoHeadingsData = [
  {
    uid: 'uid',
    label: 'UID',
    type: filterTypes.uid,
  },
  {
    uid: 'name',
    label: 'Text',
    type: filterTypes.text,
    required: true,
  },
  {
    uid: 'longText',
    label: 'Long Text',
    type: filterTypes.textLong,
  },
  {
    uid: 'date',
    label: 'Date',
    type: filterTypes.date,
  },
  {
    uid: 'number',
    label: 'Number',
    type: filterTypes.number,
  },
  {
    uid: 'number-capped',
    label: 'Number Capped',
    type: filterTypes.number,
    min: 3,
    max: 8,
    step: 0.1,
    defaultValue: 5,
  },
  {
    uid: 'demoField',
    label: 'Custom Field Types',
    type: 'demoFieldType',
  },
  {
    uid: 'reference',
    label: 'Reference',
    collection: 'democollection',
    type: filterTypes.reference,
    objectLink: true,
  },
  {
    uid: 'file',
    label: 'File',
    collectionId: 'democollection',
    documentId: 'docid',
    fileType: 'document',
    type: filterTypes.file,
  },
  {
    uid: 'image',
    label: 'Image',
    type: filterTypes.image,
  },
  {
    uid: 'bool',
    label: 'Bool',
    type: filterTypes.bool,
  },
  {
    uid: 'select',
    label: 'Select',
    type: filterTypes.select,
    options: [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectreadonly',
    label: 'Select Read Only',
    type: filterTypes.select,
    readOnly: true,
    options: [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectSearch',
    label: 'Select Search',
    type: filterTypes.selectSearch,
    searchFieldTargetField: 'label',
    searchFn: async () => [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectSearchMulti',
    label: 'Select Search Multi',
    type: filterTypes.selectSearch,
    multiple: true,
    searchFieldTargetField: 'label',
    allowEmptySearch: true,
    searchFn: async () => [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'multiSelect',
    label: 'Multi Select',
    type: filterTypes.selectMulti,
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'multiSelectlist',
    label: 'Multi Select No Dropdown',
    type: filterTypes.selectMulti,
    asDropdown: false,
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'multiSelectlistShowAll',
    label: 'Multi Select Show All',
    type: filterTypes.selectMulti,
    selectType: 'showall',
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'embedded',
    label: 'Embedded',
    type: filterTypes.embedded,
    children: [{ uid: 'embeddedtext', label: 'Embedded Text', type: filterTypes.text }],
  },
  {
    uid: 'embeddedb',
    label: 'Embedded B',
    type: filterTypes.embedded,
    orientation: 'horiz',
    children: [
      { uid: 'embeddedtog1', label: 'Embedded Tog1', type: filterTypes.bool },
      { uid: 'embeddedtog2', label: 'Embedded Tog2', type: filterTypes.bool },
      { uid: 'embeddedtog3', label: 'Embedded Tog3', type: filterTypes.bool },
    ],
  },
  {
    uid: 'dict',
    label: 'Dictionary Field',
    type: filterTypes.dict,
  },
  {
    uid: 'video',
    label: 'Video     Field',
    type: filterTypes.video,
  }
];

export const demoFormData = {
  name: 'Name 1',
  uid: 'name-1',
  date: '2019-11-02T12:04:44.626+00:00',
  selectreadonly: 'rep1',
  multiSelect: ['foo', 'bar'],
  longText: `Long Text spanning multiple lines
  1
  2
  3
  4
  5
  6
  7
  8
  9
  `,
};

export const demoAdditionalData = {
  customFieldStyle: { background: 'grey', padding: '2px' },
};
