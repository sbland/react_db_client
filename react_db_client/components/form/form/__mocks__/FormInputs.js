import React from 'react';
import PropTypes from 'prop-types';


const FormInputs = ({
  headings,
  formData,
  updateFormData,
  orientation,
  heading,
  isSection,
}) => {
  const className = `
    form_inputs
    ${(isSection) ? 'formSection' : ''}
    ${orientation}
  `;
  return (
    <section className={className}>
      {(!isSection) && <p>* is required.  (!) has been Overriden.</p>}
      <h4 className="formSection_heading">{heading}</h4>
      <div className="form_row">
        <input type="text" />
      </div>
    </section>
  );
};

FormInputs.propTypes = {
  headings: PropTypes.arrayOf(PropTypes.shape({
    uid: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    type: PropTypes.oneOf([
      'uid',
      'reference',
      'file',
      'fileMultiple',
      'textLong',
      'text',
      'number',
      'date',
      'image',
      'bool',
      'select',
      'selectMulti',
      'embedded',
    ]).isRequired,
    unit: PropTypes.string,
  })).isRequired,
  formData: PropTypes.object.isRequired,
  updateFormData: PropTypes.func.isRequired,
  orientation: PropTypes.oneOf(['horiz', 'vert']),
  heading: PropTypes.string,
  isSection: PropTypes.bool,
};

FormInputs.defaultProps = {
  orientation: 'vert',
  heading: '',
  isSection: false,
};

export default FormInputs;
