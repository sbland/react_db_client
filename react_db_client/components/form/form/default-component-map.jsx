import React from 'react';
import { filterTypes } from '@samnbuk/react_db_client.constants.client-types';

import { FieldText } from '@samnbuk/react_db_client.components.form.form-fields.field-text';
import { FieldTextArea } from '@samnbuk/react_db_client.components.form.form-fields.field-text-area';
import { FieldBool } from '@samnbuk/react_db_client.components.form.form-fields.field-bool';
import { FieldDate } from '@samnbuk/react_db_client.components.form.form-fields.field-date';
import { FieldNumber } from '@samnbuk/react_db_client.components.form.form-fields.field-number';
import { FieldObjectRef } from '@samnbuk/react_db_client.components.form.form-fields.field-object-ref';
import { FieldMultiSelect } from '@samnbuk/react_db_client.components.form.form-fields.field-multi-select';
import { FieldSelect } from '@samnbuk/react_db_client.components.form.form-fields.field-select';
import { FieldSelectSearch } from '@samnbuk/react_db_client.components.form.form-fields.field-select-search';
import { FieldFile } from '@samnbuk/react_db_client.components.form.form-fields.field-file';

export const defaultComponentMap = ({ asyncGetDocuments, fileServerUrl } = {}) => ({
  [filterTypes.text]: () => FieldText,
  [filterTypes.select]: () => FieldSelect,
  [filterTypes.selectMulti]: () => FieldMultiSelect,
  // [filterTypes.fileMultiple]: () => FieldFile,
  [filterTypes.file]: () => (props) =>
    <FieldFile {...props} fileServerUrl={fileServerUrl} asyncGetDocuments={asyncGetDocuments} />,
  // [filterTypes.image]: () => FieldFile,
  [filterTypes.textLong]: () => FieldTextArea,
  [filterTypes.number]: () => FieldNumber,
  [filterTypes.date]: () => FieldDate,
  [filterTypes.bool]: () => FieldBool,
  [filterTypes.toggle]: () => FieldBool,
  [filterTypes.selectSearch]: () => FieldSelectSearch,
  [filterTypes.reference]: () => (props) =>
    <FieldObjectRef {...props} asyncGetDocuments={asyncGetDocuments} />,
});
