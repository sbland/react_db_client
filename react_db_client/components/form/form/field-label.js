import React from 'react';
import PropTypes from 'prop-types';

export const FieldLabel = ({
  label,
  inputClassName,
  hasChanged,
  required,
  hasLabel,
}) => (
  <label className={inputClassName}>
    {required && '*'}
    {hasLabel && (
      <>
        {label}
        {/* Below is a possible hack to stop chrome autofill */}
        {/* {label.split('').join('\u200b')} */}
        {hasChanged && '(!)'}
        {':'}{' '}
      </>
    )}
    {!hasLabel && <span />}
  </label>
);

FieldLabel.propTypes = {
  label: PropTypes.string.isRequired,
  inputClassName: PropTypes.string,
  hasChanged: PropTypes.bool,
  required: PropTypes.bool,
  hasLabel: PropTypes.bool,
};

FieldLabel.defaultProps = {
  inputClassName: '',
  hasChanged: false,
  required: false,
  hasLabel: false,
};
