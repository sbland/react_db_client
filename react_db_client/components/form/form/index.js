export { Form } from './form';
export { FormInputs } from './FormInputs';
export * as FormStyles from './_form.scss';
export { defaultComponentMap } from './default-component-map';
export { FormField } from './FormField';