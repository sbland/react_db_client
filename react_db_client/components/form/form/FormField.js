import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { filterTypes } from '@react_db_client/constants.client-types';
import { switchF } from '@react_db_client/helpers.func-tools';

import { FieldReadOnly } from '@samnbuk/react_db_client.components.form.form-fields.field-read-only';
import { FieldLabel } from './field-label';

import './_form.scss';

const defaultComponent = () => FieldReadOnly;

export const FormField = (propsIn) => {
  const { heading, value, updateFormData, additionalData, componentMap } = propsIn;
  const props = useMemo(
    () => ({
      updateFormData,
      value,
      key: `${heading.uid}-sub`,
      additionalData,
      ...heading,
    }),
    [heading, updateFormData, value, additionalData]
  );
  const { label, required, type, uid, hasChanged, readOnly } = heading;

  const FormComponent = useMemo(
    () => switchF(heading.type, componentMap, defaultComponent),
    [heading.type, componentMap, defaultComponent]
  );

  const labelClassName = [
    'form_label',
    `${required ? 'required' : ''}`,
    `${hasChanged ? 'hasChanged' : ''}`,
  ]
    .filter((f) => f)
    .join(' ');
  const rowClassname = ['form_row', `form_row_heading_id_${uid}`].filter((f) => f).join(' ');

  const showLabel = [filterTypes.bool, filterTypes.toggle].indexOf(type) === -1 || heading.readOnly; // we do not need a label for a toggle box
  return (
    <div className={rowClassname} key={uid}>
      <FieldLabel
        uid={uid}
        label={label}
        inputClassName={labelClassName}
        hasChanged={hasChanged}
        required={required}
        hasLabel={showLabel}
      />
      <FormComponent {...props} />
    </div>
  );
};

FormField.propTypes = {
  heading: PropTypes.shape({
    uid: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    unit: PropTypes.string,
    required: PropTypes.bool,
    hasChanged: PropTypes.bool,
    readOnly: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['true', 'false'])]),
  }).isRequired,
  value: PropTypes.any,
  updateFormData: PropTypes.func.isRequired,
  additionalData: PropTypes.shape({}).isRequired,
  componentMap: PropTypes.objectOf(PropTypes.elementType),
};

FormField.defaultProps = {
  value: null,
  componentMap: {},
};
