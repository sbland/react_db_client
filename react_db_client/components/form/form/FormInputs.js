import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { filterTypes } from '@react_db_client/constants.client-types';

import '@samnbuk/react_db_client.constants.style';
import { FormField as DefaultFormField } from './FormField';
import './_form.scss';

export const FormInputs = ({
  FormField,
  headings,
  formData,
  updateFormData,
  orientation,
  heading: sectionHeading,
  isSection,
  showKey,
  additionalData,
  componentMap,
  id,
}) => {
  const className = [
    'form_inputs',
    'formSection',
    `${orientation}`,
    sectionHeading ? 'hasHeading' : '',
    `formSection_${id}`,
  ]
    .filter((f) => f)
    .join(' ');

  const fields = useMemo(
    () =>
      headings.map((heading) => {
        const value =
          formData && formData[heading.uid] !== null
            ? formData[heading.uid]
            : heading.defaultValue || heading.def;

        if (heading.type === filterTypes.embedded) {
          return (
            <FormInputs
              key={heading.uid}
              headings={heading.children}
              formData={formData}
              updateFormData={updateFormData}
              orientation={heading.orientation}
              heading={heading.label}
              additionalData={additionalData}
              componentMap={componentMap}
              FormField={FormField}
              id={heading.uid}
              isSection
            />
          );
        }
        // return <div>{heading.uid}</div>;
        return (
          <FormField
            heading={heading}
            updateFormData={updateFormData}
            value={value}
            key={heading.uid}
            additionalData={additionalData}
            componentMap={componentMap}
          />
        );
      }),
    [headings, formData, updateFormData, additionalData, componentMap]
  );

  return (
    <section className={className}>
      {!isSection && showKey && <p>* is required. (!) has been modified.</p>}
      <h4 className="formSection_heading">{sectionHeading}</h4>
      {fields}
    </section>
  );
};

FormInputs.propTypes = {
  FormField: PropTypes.func,
  headings: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      orientation: PropTypes.oneOf(['horiz', 'vert']),
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      unit: PropTypes.string,
      children: PropTypes.arrayOf(PropTypes.shape({})),
    })
  ).isRequired,
  formData: PropTypes.object.isRequired,
  updateFormData: PropTypes.func.isRequired,
  orientation: PropTypes.oneOf(['horiz', 'vert']),
  heading: PropTypes.string,
  isSection: PropTypes.bool,
  showKey: PropTypes.bool,
  additionalData: PropTypes.shape({}),
  componentMap: PropTypes.objectOf(PropTypes.elementType).isRequired,
  id: PropTypes.any,
};

FormInputs.defaultProps = {
  FormField: DefaultFormField,
  orientation: 'vert',
  heading: '',
  isSection: false,
  showKey: true,
  additionalData: {},
  id: null,
};
