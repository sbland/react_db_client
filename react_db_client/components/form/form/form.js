import React, { useCallback, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';

import cloneDeep from 'lodash/cloneDeep';

import { FormInputs } from './FormInputs';

import './_form.scss';
import formValidation from './formValidation';
// import Emoji from '../Emoji/Emoji';

/** Generic Form Component that manages updates and layout
 *
 * The Form component handles general form requirements such as storing
 * edit data, auto generating a layout bassed on provided headers.


 # Headings
 The headings should have the correct headings shape...

 # Mounting the save buttons using a ref:
 If a react ref is provided to `endButtonRefOverride` then the save buttons will be placed in
 the referenced container.


 * On submit signature:
 * ({ formEditData, formData }) => {...}
 *
 * onChange Signature:
 * (field, value, newFormData) => {...}


 * additionalData - additional data to be passed to form field components. Useful when using a custom field
 * customFieldComponents - a map of field type against react component
 */
export const Form = ({
  FormField,
  formDataInitial,
  headings,
  onSubmit,
  onChange,
  showEndBtns,
  submitBtnText,
  showKey,
  orientation,
  disableAutocomplete,
  endButtonRefOverride,
  errorCallback,
  additionalData,
  customFieldComponents,
}) => {
  // const [formData, setFormData] = useState(cloneDeep(formDataInitial));
  const [formEditData, setFormEditData] = useState({});
  const [endButtonContainerRef, setEndButtonContainerRef] = useState(null);

  const formData = useMemo(() => ({ ...cloneDeep(formDataInitial), ...formEditData }), [
    formDataInitial,
    formEditData,
  ]);

  const updateFormData = useCallback(
    (field, value) => {
      setFormEditData((prev) => {
        const newFormData = cloneDeep(prev);
        newFormData[field] = value;
        if (onChange) onChange(field, value, newFormData);
        return newFormData;
      });
    },
    [onChange]
  );

  const handleReset = () => {
    setFormEditData({});
  };

  const handleSubmit = useCallback(() => {
    const passesFormValidation = formValidation(formData, headings);
    if (passesFormValidation === true) onSubmit({ formEditData, formData });
    else errorCallback(passesFormValidation.error);
  }, [formData, formEditData, headings, onSubmit, errorCallback]);

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        handleSubmit();
      }}
      // Another possible hack to stop autocomplete on chrome
      // autoComplete={disableAutocomplete ? 'chrome-off' : 'on'}
      autoComplete={disableAutocomplete ? 'off' : 'on'}
      className="form sectionWrapper"
    >
      <FormInputs
        headings={headings}
        formData={formData}
        updateFormData={updateFormData}
        showKey={showKey}
        orientation={orientation}
        additionalData={additionalData}
        customFieldComponents={customFieldComponents}
        FormField={FormField}
      />
      <section ref={(ref) => setEndButtonContainerRef(ref)} style={{ width: '100%' }} />
      {showEndBtns &&
        endButtonContainerRef &&
        ReactDOM.createPortal(
          <div className="submitBtns">
            {/* TODO: Work out why setting this as a submit button does not work */}
            <button type="button" className="button-two submitBtn" onClick={handleSubmit}>
              <span role="img" aria-label={submitBtnText}>
                üíæ
              </span>
            </button>
            <button type="button" className="button-one" onClick={handleReset}>
              <span role="img" aria-label="Delete">
                ‚ùå
              </span>
            </button>
          </div>,
          endButtonRefOverride || endButtonContainerRef
        )}
    </form>
  );
};

Form.propTypes = {
  formDataInitial: PropTypes.shape().isRequired,
  headings: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.oneOf([
        'uid',
        'reference',
        'file',
        'textLong',
        'text',
        'number',
        'date',
        'image',
        'bool',
        'select',
        'selectMulti',
        'embedded',
      ]).isRequired,
    })
  ).isRequired,
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func,
  showEndBtns: PropTypes.bool,
  submitBtnText: PropTypes.string,
  disableAutocomplete: PropTypes.bool,
  showKey: PropTypes.bool,
  orientation: PropTypes.oneOf(['horiz', 'vert']),
  endButtonRefOverride: PropTypes.any, // Should be a react reference
  errorCallback: PropTypes.func,
  additionalData: PropTypes.shape({}),
  customFieldComponents: PropTypes.objectOf(PropTypes.elementType),
};

Form.defaultProps = {
  onChange: () => {},
  showEndBtns: true,
  submitBtnText: 'Save',
  showKey: true,
  orientation: 'vert',
  disableAutocomplete: false,
  endButtonRefOverride: null,
  errorCallback: alert,
  additionalData: {},
  customFieldComponents: {},
};

