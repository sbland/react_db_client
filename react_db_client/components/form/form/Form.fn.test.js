import '@samnbuk/react_db_client.helpers.enzyme-setup';
import React from 'react';
import { mount } from 'enzyme';
import { SearchAndSelectDropdown } from '@samnbuk/react_db_client.components.search-and-select-dropdown';

import { demoHeadingsData, demoFormData } from './DemoData';

import { Form } from './form';
import { FieldText } from '@samnbuk/react_db_client.components.form.form-fields.field-text';
import { FieldReadOnly } from '@samnbuk/react_db_client.components.form.form-fields.field-read-only';
import { FieldObjectRef } from '@samnbuk/react_db_client.components.form.form-fields.field-object-ref';
import { FormInputs } from './FormInputs';
import { FieldLabel } from './field-label';
import { FormField } from './FormField';
import { filterTypes } from '@samnbuk/react_db_client.constants.client-types';
import { defaultComponentMap } from './default-component-map';

const onSubmit = jest.fn();
const onChange = jest.fn();

const defaultProps = {
  headings: demoHeadingsData,
  formDataInitial: demoFormData,
  onSubmit,
  onChange,
  showEndBtns: true,
  submitBtnText: 'Submit',
  showKey: true,
  orientation: 'vert',
  additionalData: { hello: 'world' },
  componentMap: defaultComponentMap,
  FormField,
};

describe('Form - Functional Tests', () => {
  beforeEach(() => {
    onChange.mockClear();
    onSubmit.mockClear();
  });
  describe('Functional tests', () => {
    let component;
    beforeEach(() => {
      component = mount(<Form {...defaultProps}
      />);
    });
    describe('Bug fixing', () => {
      // TODO: Below no longer works but component is working
      test.skip('should call on change when a reference dropdown is selected', () => {
        const formFieldObjectRef = component.find(FieldObjectRef);
        expect(formFieldObjectRef.props()).toEqual({
          ...demoHeadingsData[6],
          labelField: 'label',
          multiple: false,
          required: false,
          unit: '',
          updateFormData: expect.any(Function),
          value: null,
        });
        const newValue = 'newval';
        const newFormData = { [demoHeadingsData[6].uid]: newValue };
        formFieldObjectRef.props().updateFormData(demoHeadingsData[6].uid, newValue);
        expect(onChange).toHaveBeenCalledWith(demoHeadingsData[6].uid, newValue, newFormData);
        const sas = formFieldObjectRef.find(SearchAndSelectDropdown);
        expect(sas.props()).toEqual({
          allowEmptySearch: true,
          allowMultiple: false,
          className: 'formFieldInput',
          handleSelect: expect.any(Function),
          labelField: 'label',
          required: false,
          returnFieldOnSelect: '_id',
          searchDelay: 500,
          searchFieldPlaceholder: null,
          searchFieldTargetField: 'label',
          searchFunction: expect.any(Function),
          selectionOverride: null,
        });
        onChange.mockClear();
        sas.props().handleSelect(newValue);
        expect(onChange).toHaveBeenCalledWith(demoHeadingsData[6].uid, newValue, newFormData);
      });
    });

    describe('Passing data to fields', () => {
      test('should pass headings data to field', () => {
        const field = component.find(FormField).first().childAt(0).childAt(1);
        expect(field.type()).toEqual(FieldReadOnly);
        expect(field.props()).toEqual({
          label: demoHeadingsData[0].label,
          type: demoHeadingsData[0].type,
          options: demoHeadingsData[0].options || [],
          uid: demoHeadingsData[0].uid,
          unit: demoHeadingsData[0].unit || '',
          updateFormData: expect.any(Function),
          value: demoFormData.uid,
          additionalData: defaultProps.additionalData,
        });
        expect(field.debug()).toMatchSnapshot();
      });
      test('should pass headings data to field label', () => {
        const field = component.find(FormField).first().childAt(0).childAt(0);
        expect(field.type()).toEqual(FieldLabel);
        expect(field.props()).toEqual({
          label: demoHeadingsData[0].label,
          required: demoHeadingsData[0].required || false,
          hasChanged: false,
          hasLabel: true,
          inputClassName: 'form_label',
        });
        expect(field.debug()).toMatchSnapshot();
      });
      test('should pass field update back to onChange function', () => {
        const field = component.find(FormField).find(FieldText).first()
        expect(field.type()).toEqual(FieldText);
        expect(field.props().uid).toEqual(demoHeadingsData[1].uid);
        const input = field.find('input');
        const newValue = 'New Val';
        const newFormData = { [demoHeadingsData[1].uid]: newValue };
        input.simulate('change', { target: { value: newValue } });
        expect(onChange).toHaveBeenCalledWith(demoHeadingsData[1].uid, newValue, newFormData);
      });
      test('should pass label value to read only select field', () => {
        const fieldIndex = 12;
        const field = component.find(FormField).at(fieldIndex).childAt(0).childAt(1);
        expect(field.type()).toEqual(FieldReadOnly);
        // TODO: This is a fragile test. Should link to headings data etc
        expect(field.text()).toEqual('Rep 1');
      });
      test('should pass additional data to fields', () => {
        // We can include additional data in the form that is accessible by each field.
        const additionalData = { hello: 'world' };
        component.setProps({ additionalData });
        component.update();
        const formInputs = component.find(FormInputs).first();
        expect(formInputs.props().additionalData).toEqual(additionalData);
        const field = component.find(FormField).first();
        expect(field.props().additionalData).toEqual(additionalData);
      });
    });
  });
});
