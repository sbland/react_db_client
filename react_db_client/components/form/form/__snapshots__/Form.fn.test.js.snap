// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form - Functional Tests Functional tests Passing data to fields should pass headings data to field 1`] = `
"<FieldReadOnly updateFormData={[Function (anonymous)]} value=\\"name-1\\" additionalData={{...}} uid=\\"uid\\" label=\\"UID\\" type=\\"uid\\" unit=\\"\\" options={{...}}>
  <span>
    name-1
  </span>
  <span className=\\"unitSpan\\" />
</FieldReadOnly>"
`;

exports[`Form - Functional Tests Functional tests Passing data to fields should pass headings data to field label 1`] = `
"<FieldLabel label=\\"UID\\" inputClassName=\\"form_label\\" hasChanged={false} required={false} hasLabel={true}>
  <label className=\\"form_label\\">
    UID
    :
     
  </label>
</FieldLabel>"
`;
